text/0000-inferred-variant-constructor.md

```diff
@@ -0,0 +1,204 @@
+- Feature Name: `inferred_variant_constructor`
+- Start Date: 2017-03-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow the name (qualifier) of an enum variant to be elided in expressions and
+patterns whenever it can be inferred.
+
+# Motivation
+[motivation]: #motivation
+
+The main goal of this proposal is to improve ergonomics of enums in Rust by
+reducing verbosity.  Currently, there are two convenient ways to use enums.
+One would either:
+
+  - bring the enum type into scope (`use foolib::Foo;`), or
+  - bring the enum variants into scope (`use foolib::Foo::*;`).
+
+The former is less convenient than the latter, but the latter pollutes the
+namespace more.  One could always write enum variants fully qualified,
+`foolib::Foo::Alpha`, but that is even more tedious.
+
+This proposal intends to make variant expressions as convenient and
+pollution-free as field expressions today.  Whereas field expressions allow
+the user to extract fields of a struct without ever bringing the struct into
+scope, inferred variant constructors allow the user to construct/match
+variants of an enum without ever bringing the enum or its variants into scope.
+
+This could lower the usability barrier for enums, particularly for library
+users.  It would reduce the verbosity of enums to a level similar to
```

Comment:
> When you have clashes, then its even more important to distinguish between them inside your code. Otherwise you lose readability.

I believe that in cases like function arguments the context provides readability. In general I agree with you though. I'm in the "import as little as possible" camp.

> This is an RFC to propose changes to the rust language, not the official rust style. If you don't like the official style, propose to change it instead.

Change it to what? I see disadvantages in both "prefix all the variants" and "import the variants" styles. I feel like a solution with the functionality like this RFC is a third, in between way that hits a sweet spot.